src/dali/abstract_pair_converter_plugin.py

```diff
@@ -67,19 +67,40 @@ class AssetPairAndTimestamp(NamedTuple):
     exchange: str
 
 
+class TransactionManifest(NamedTuple):
+    assets: Set[str]
+    exchanges: Set[str]
+    first_transaction_datetime: datetime
+
+
 class MappedGraph(Graph[ValueType]):
-    def __init__(self, vertexes: Optional[List["Vertex[ValueType]"]] = None) -> None:
+    def __init__(self, vertexes: Optional[List["Vertex[ValueType]"]] = None, optimized_assets: Optional[Set[str]] = None) -> None:
         super().__init__(vertexes)
         self.__name_to_vertex: Dict[str, Vertex[ValueType]] = {vertex.name: vertex for vertex in vertexes} if vertexes else {}
+        self.__optimized_assets: Set[str] = set() if optimized_assets is None else optimized_assets
+
+    def add_vertex_if_missing(self, name: str) -> None:
+        if not self.__name_to_vertex.get(name):
+            self.add_vertex(Vertex[ValueType](name=name))
+
+    def get_all_children_of_vertex(self, vertex: Vertex[ValueType]) -> Set[Vertex[ValueType]]:
```

Comment:
The graphs being built don't loop back around. They only move toward the quote asset. Having said that though, we should include some kind of safety since this could infinitely recurse under some unforeseen situation.

Reply:
Agreed, we should throw an internal error if a cycle is detected.

Reply:
This does actually lead to infinite recursion with real-world data. I'll have to rewrite it to prevent that.